### PR TITLE
Update `block-indentation` rule to add new `ignoreComments` option

### PR DESCRIPTION
**Pull request by akashdsouza**
For https://github.com/ember-template-lint/ember-template-lint/issues/514

I wasn't sure if this was the exact requirement. Let me know if any changes are to be made

### CODE REVIEW

File: lib/rules/block-indentation.js [hunk 1]
```diff
@@ -83,16 +85,46 @@ module.exports = class BlockIndentation extends Rule {
...
-        return config;
+        return {
+          indentation: config,
+        };
       case 'boolean':
-        return config ? 2 : false;
+        if (config) {
+          return {
+            indentation: 2,
+          };
```
Comment (bmish) [inline lib/rules/block-indentation.js]: Can you store this in a `DEFAULT_CONFIG` object?

File: lib/rules/block-indentation.js [hunk 2]
```diff
@@ -45,7 +45,9 @@
  * boolean -- `true` indicates a 2 space indent, `false` indicates that the rule is disabled.
  * numeric -- the number of spaces to require for indentation
  * "tab" -- To indicate tab style indentation (1 char)
-
+ * object --
+ *    `indentation: <numeric>` - number of spaces to indent',
```
Comment (bmish) [inline lib/rules/block-indentation.js]: Can you mention add `(default 2)` to this line?

File: lib/rules/block-indentation.js [hunk 3]
```diff
@@ -45,7 +45,9 @@
...
+      `skip-comments: <boolean>` - skip indentation for comments',
```
Comment (bmish) [inline lib/rules/block-indentation.js]: Can you add `(default false)` to this line?

File: lib/rules/block-indentation.js [hunk 4]
```diff
@@ -83,16 +85,46 @@ module.exports = class BlockIndentation extends Rule {
...
+          if (typeof skipComments === 'boolean') {
+            result.skipComments = skipComments;
+          }
+        }
+        if ('indentation' in config) {
+          let indentation = config.indentation;
+          if (typeof indentation === 'number') {
+            result.indentation = indentation;
+          } else {
+            break;
```
Comment (bmish) [inline lib/rules/block-indentation.js]: Can you add asserts for unexpected config input?
```
const assert = require('assert');

assert(false, 'unexpected config indentation');
```

File: lib/rules/block-indentation.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (akashdsouza) [reply lib/rules/block-indentation.js]: Updated

File: test/unit/rules/block-indentation-test.js [hunk 5]
```diff
@@ -116,6 +116,18 @@ generateRuleTests({
...
+        'skip-comments': true,
+      },
+      template: ['<div>', '<!-- Comment -->', '{{! Comment }}', '</div>'].join('\n'),
+    },
+    {
+      config: {
+        'skip-comments': false,
+      },
+      template: ['<div>', '  <!-- Comment -->', '  {{! Comment }}', '</div>'].join('\n'),
+    },
```
[the remark this answers is not in the document]
Comment (akashdsouza) [reply test/unit/rules/block-indentation-test.js]: I have added a couple of tests. Let me know if more are needed.

File: lib/rules/block-indentation.js (hunk 1 above)
Comment (akashdsouza) [reply lib/rules/block-indentation.js]: Done. I wasn't sure if `ignoreComments` should also be specified in the `DEFAULT_CONFIG`

File: lib/rules/block-indentation.js (hunk 2 above)
Comment (akashdsouza) [reply lib/rules/block-indentation.js]: Updated

File: lib/rules/block-indentation.js (hunk 4 above)
Comment (akashdsouza) [reply lib/rules/block-indentation.js]: Done

File: lib/rules/block-indentation.js [hunk 6]
```diff
@@ -83,16 +90,43 @@ module.exports = class BlockIndentation extends Rule {
...
+          };
         }
         break;
+      case 'object': {
+        let result = DEFAULT_CONFIG;
+        if ('ignoreComments' in config) {
+          let ignoreComments = config.ignoreComments;
+          if (typeof ignoreComments === 'boolean') {
+            result.ignoreComments = ignoreComments;
+          }
```
Comment (bmish) [inline lib/rules/block-indentation.js]: There should also be an assert here:

```
else {
  assert(false, '`ignoreComments` should be a boolean`);
}
```

File: lib/rules/block-indentation.js [hunk 7]
```diff
@@ -101,6 +135,9 @@ module.exports = class BlockIndentation extends Rule {
         '  * boolean - `true` to enable 2 space indentation',
         '  * numeric - the number of spaces to require',
         '  * "tab" - usage of one character indentation (tab char)',
+        '  * object - An object with the following keys:',
+        '    * `indentation: <numeric>` - number of spaces to indent',
+        '    * `ignoreComments: <boolean>` - skip comment indentation',
```
Comment (bmish) [inline lib/rules/block-indentation.js]: This should match the comment in `block-indentation.js` (should also mention the default values here).

And the rule documentation file needs to be updated to mention this new option too.

File: test/unit/rules/block-indentation-test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bmish) [reply test/unit/rules/block-indentation-test.js]: I would still like to see tests added for single lines that include both code and comments in various positions inside the line:

```
{{! Comment }}<p>some text</p>
```

```
<p>{{! Comment }}</p>
```

```
<p>some text</p>{{! Comment }}
```

[the remark this answers is not in the document]
Comment (bmish) [reply test/unit/rules/block-indentation-test.js]: I would still like to see tests added for single lines that include both code and comments in various positions inside the line:

```
{{! Comment }}<p>some text</p>
```

```
<p>{{! Comment }}</p>
```

```
<p>some text</p>{{! Comment }}
```

[the remark this answers is not in the document]
Comment (akashdsouza) [reply test/unit/rules/block-indentation-test.js]: Sorry, I didn't read the first comment properly. I have updated the tests accordingly

**Comment by akashdsouza** [comment]
I have made the required changes. @bmish let me know if anything else needs to be updated

File: test/unit/rules/block-indentation-test.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bmish) [reply test/unit/rules/block-indentation-test.js]: Thanks, also I added a few more for you.

File: lib/rules/block-indentation.js [hunk 8]
```diff
@@ -83,16 +90,48 @@ module.exports = class BlockIndentation extends Rule {
...
+            indentation: 1,
+          };
         }
         break;
+      case 'object': {
+        let result = DEFAULT_CONFIG;
+        if ('ignoreComments' in config) {
+          let ignoreComments = config.ignoreComments;
+          if (typeof ignoreComments === 'boolean') {
+            result.ignoreComments = ignoreComments;
```
Comment (rwjblue) [inline lib/rules/block-indentation.js]: This is mutating the shared constant `DEFAULT_CONFIG`

Comment (rwjblue) [reply lib/rules/block-indentation.js]: Fixing in https://github.com/ember-template-lint/ember-template-lint/pull/1144

**Pull request merged**